### PR TITLE
Allow providing external HA1 for digest auth

### PR DESCRIPTION
**Pull request by manuelottlik**
This PR implements #332. It adds nothing but optional parameters for `ha1` and its documentation.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,8 +18,8 @@ Version 5 is under active development. Version 4 is in support mode, and will re
 
 Version 5 upgrades the library to use ESM (ECMAScript Modules), and so your environment must fit one of the following formats to be able to use this library:
 
- * NodeJS project with `"type": "module"` in `package.json` (ESM mode)
- * Web project bundled with a tool like Webpack that can handle ESM
+-   NodeJS project with `"type": "module"` in `package.json` (ESM mode)
+-   Web project bundled with a tool like Webpack that can handle ESM
```
Comment (perry-mitchell) [inline README.md:22]: Any particular reason you change these? `*` are perfectly valid markdown, as are the `|---|` (no spaces) for tables.

File: README.md [hunk 2]
```diff
@@ -169,17 +171,17 @@ The `createClient` method takes a WebDAV service URL, and a configuration option
...
+| `password`        | _None_                                                                                    | Password for authentication.                                                                                                              |
+| `token`           | _None_                                                                                    | Token object for authentication.                                                                                                          |
+| `username`        | _None_                                                                                    | Username for authentication.                                                                                                              |
+| `withCredentials` | _None_                                                                                    | Credentials inclusion setting for the request,                                                                                            |
```
Comment (perry-mitchell) [inline README.md:184]: This is actually harder to read in a width-constricted terminal, when compared to the shortened version that was there before. Are you using some markdown formatting utility perhaps?

File: README.md [hunk 3]
```diff
@@ -239,33 +238,29 @@ Synchronously create a readable stream for a remote file.
 _Note that although a stream is returned instantly, the connection and fetching of the file is still performed asynchronously in the background. There will be some delay before the stream begins receiving data._
 
 ```typescript
-client
-    .createReadStream("/video.mp4")
-    .pipe(fs.createWriteStream("~/video.mp4"));
+client.createReadStream("/video.mp4").pipe(fs.createWriteStream("~/video.mp4"));
```
Comment (perry-mitchell) [inline README.md:241]: I prefer the stacked format to this..

File: README.md [hunk 4]
```diff
@@ -276,21 +271,20 @@ Create a write stream targeted at a remote file.
 _Note that although a stream is returned instantly, the connection and writing to the remote file is still performed asynchronously in the background. There will be some delay before the stream begins piping data._
 
 ```typescript
-fs
-    .createReadStream("~/Music/song.mp3")
-    .pipe(client.createWriteStream("/music/song.mp3"));
+fs.createReadStream("~/Music/song.mp3").pipe(client.createWriteStream("/music/song.mp3"));
```
Comment (perry-mitchell) [inline README.md:274]: I prefer the original format

File: source/auth/digest.ts [hunk 5]
```diff
@@ -5,8 +5,12 @@ import { DigestContext } from "../types.js";
...
 const NONCE_SIZE = 32;
 
-export function createDigestContext(username: string, password: string): DigestContext {
-    return { username, password, nc: 0, algorithm: "md5", hasDigestAuth: false };
+export function createDigestContext(
+    username: string,
+    password: string,
+    ha1: string
+): DigestContext {
+    return { username, password, ha1, nc: 0, algorithm: "md5", hasDigestAuth: false };
```
Comment (perry-mitchell) [inline source/auth/digest.ts:13]: Would you mind merging/rebasing master? I added formatting (prettier) tests.. want to make sure these changes still pass.

File: source/tools/crypto.ts [hunk 6]
```diff
@@ -6,9 +6,11 @@ export function ha1Compute(
     realm: string,
     pass: string,
     nonce: string,
-    cnonce: string
+    cnonce: string,
+    ha1: string
 ): string {
-    const ha1 = md5(`${user}:${realm}:${pass}`) as string;
+    ha1 = ha1 || (md5(`${user}:${realm}:${pass}`) as string);
+
```
Comment (perry-mitchell) [inline source/tools/crypto.ts:13]: No need for blank line

File: source/types.ts [hunk 7]
```diff
@@ -257,6 +258,7 @@ export interface WebDAVClientContext {
     token?: OAuthToken;
     username?: string;
     withCredentials?: boolean;
+    ha1?: string;
```
Comment (perry-mitchell) [inline source/types.ts:261]: Would you mind sorting this alphabetically?

File: source/types.ts [hunk 8]
```diff
@@ -276,6 +278,7 @@ export interface WebDAVClientOptions {
...
```
Comment (perry-mitchell) [inline source/types.ts:281]: Would you mind sorting this alphabetically?

File: source/factory.ts [hunk 9]
```diff
@@ -47,7 +47,8 @@ export function createClient(remoteURL: string, options: WebDAVClientOptions = {
         password,
         token,
         username,
-        withCredentials
+        withCredentials,
+        ha1
```
Comment (perry-mitchell) [inline source/factory.ts:51]: Would you mind sorting this alphabetically?

File: source/factory.ts [hunk 10]
```diff
@@ -64,9 +65,10 @@ export function createClient(remoteURL: string, options: WebDAVClientOptions = {
...
```
Comment (perry-mitchell) [inline source/factory.ts:69]: Would you mind sorting this alphabetically?

**Review by perry-mitchell — CHANGES_REQUESTED** [review]
Looks awesome! Just a few items to work out. There's a lot of changes so I'd be surprised if there weren't a few :)

Thanks for this!

**Comment by manuelottlik** [comment]
Hey @perry-mitchell, thank you for the review. All the formatting stuff was not intentional, just some extension – should have payed attention to that, sorry. I will fix all of this, rebase and also order all arguments alphabetically.

**Comment by manuelottlik** [comment]
Hey @perry-mitchell, I removed all the re-formatting, merged your master into the branch and ordered all arguments alphabetically. Is there anything else I can do?

**Comment by perry-mitchell** [comment]
I'll check this shortly, thank you!

**Review by perry-mitchell — APPROVED** [review]

**Pull request merged**